File: java/src/main/java/vsdk/toolkit/tonemap/WardToneMap.java
package vsdk.toolkit.tonemap;

import vsdk.toolkit.common.Cie;
import vsdk.toolkit.common.ColorRgb;

/**
References:

G. Ward. A Contrast-Based Scale factor for Luminance Display, Graphics
Gems IV, Academic Press, 1994, pp. 415-421.
*/
public final class WardToneMap extends ToneMap {
    private float comp;
    private float display;
    private float lda;

    public WardToneMap() {
        comp = 0.0f;
        display = 0.0f;
        lda = 0.0f;
    }

    @Override
    public void init(ToneMappingContext toneMapOptions) {
        float realWorldAdaptionLuminance = toneMapOptions.realWorldAdaptionLuminance;
        float maximumDisplayLuminance = toneMapOptions.maximumDisplayLuminance;
        lda = maximumDisplayLuminance / 2.0f;

        float p1 = (float)Math.pow(lda, 0.4);
        float p2 = (float)Math.pow(realWorldAdaptionLuminance, 0.4);
        float p3 = (1.219f + p1) / (1.219f + p2);
        comp = (float)Math.pow(p3, 2.5);
        display = comp / maximumDisplayLuminance;
    }

    @Override
    public ColorRgb scaleForComputations(ColorRgb radiance) {
        radiance.scale(comp);
        return radiance;
    }

    @Override
    public ColorRgb scaleForDisplay(ColorRgb radiance) {
        float eff = Cie.getLuminousEfficacy();

        radiance.scale(eff * display);
        return radiance;
    }
}

File: typescript/src/main/vsdk/toolkit/tonemap/WardToneMap.ts
import { Cie } from "../common/Cie";
import { ColorRgb } from "../common/ColorRgb";
import { ToneMap } from "./ToneMap";
import { ToneMappingContext } from "./ToneMappingContext";

export class WardToneMap extends ToneMap {
  private comp: number;
  private display: number;
  private lda: number;

  public constructor() {
    super();
    this.comp = 0.0;
    this.display = 0.0;
    this.lda = 0.0;
  }

  public override init(toneMapOptions: ToneMappingContext): void {
    const realWorldAdaptionLuminance = toneMapOptions.realWorldAdaptionLuminance;
    const maximumDisplayLuminance = toneMapOptions.maximumDisplayLuminance;
    this.lda = maximumDisplayLuminance / 2.0;

    const p1 = globalThis.Math.pow(this.lda, 0.4);
    const p2 = globalThis.Math.pow(realWorldAdaptionLuminance, 0.4);
    const p3 = (1.219 + p1) / (1.219 + p2);
    this.comp = globalThis.Math.pow(p3, 2.5);
    this.display = this.comp / maximumDisplayLuminance;
  }

  public override scaleForComputations(radiance: ColorRgb): ColorRgb {
    radiance.scale(this.comp);
    return radiance;
  }

  public override scaleForDisplay(radiance: ColorRgb): ColorRgb {
    const eff = Cie.getLuminousEfficacy();
    radiance.scale(eff * this.display);
    return radiance;
  }
}
